packages/happy-dom/src/xml-http-request/XMLHttpRequest.ts

```diff
@@ -0,0 +1,655 @@
+			this.readyState > XMLHttpRequestReadyStateEnum.opened &&
+			this._response.headers[lowerHeader] &&
+			!this._errorFlag
+		) {
+			return this._response.headers[lowerHeader];
+		}
+
+		return null;
+	}
+
+	/**
+	 * Gets all the response headers.
+	 *
+	 * @returns A string with all response headers separated by CR+LF.
+	 */
+	public getAllResponseHeaders(): string {
+		if (this.readyState < XMLHttpRequestReadyStateEnum.headersRecieved || this._errorFlag) {
+			return '';
+		}
+		let result = '';
+
+		for (const name of Object.keys(this._response.headers)) {
+			// Cookie headers are excluded
+			if (name !== 'set-cookie' && name !== 'set-cookie2') {
+				result += `${name}: ${this._response.headers[name]}\r\n`;
+			}
+		}
+
+		return result.substr(0, result.length - 2);
+	}
+
+	/**
+	 * Gets a request header
+	 *
+	 * @param name Name of header to get.
+	 * @returns Returns the request header or empty string if not set.
+	 */
+	public getRequestHeader(name: string): string {
+		const lowerName = name.toLowerCase();
+		if (typeof name === 'string' && this._requestHeaders[lowerName]) {
+			return this._requestHeaders[lowerName];
+		}
+
+		return '';
+	}
+
+	/**
+	 * Sends the request to the server.
+	 *
+	 * @param data Optional data to send as request body.
+	 */
+	public send(data?: string): void {
+		if (this.readyState != XMLHttpRequestReadyStateEnum.opened) {
+			throw new Error('INVALID_STATE_ERR: connection must be opened before send() is called');
+		}
+
+		if (this._sendFlag) {
+			throw new Error('INVALID_STATE_ERR: send has already been called');
+		}
+
+		let ssl = false;
+		let local = false;
+		const url = RelativeURL.getAbsoluteURL(
+			this._ownerDocument.defaultView.location,
+			this._settings.url
+		);
+		let host;
+
+		// Determine the server
+		switch (url.protocol) {
+			case 'https:':
+				host = url.hostname;
+				ssl = true;
+				break;
+
+			case 'http:':
+				host = url.hostname;
+				break;
+
+			case 'file:':
+				local = true;
+				break;
+
+			case undefined:
+			case '':
+				host = 'localhost';
+				break;
+
+			default:
+				throw new Error('Protocol not supported.');
+		}
+
+		// TODO: Security Issue.
+		// Load files off the local filesystem (file://)
+		if (local) {
+			if (this._settings.method !== 'GET') {
+				throw new Error('XMLHttpRequest: Only GET method is supported');
+			}
+
+			if (this._settings.async) {
+				FS.readFile(decodeURI(url.pathname.slice(1)), 'utf8', (error: Error, data: Buffer) => {
+					if (error) {
+						this._handleError(error);
+					} else {
+						this.status = 200;
+						this.responseText = data.toString();
+						this._setState(XMLHttpRequestReadyStateEnum.done);
+					}
+				});
+			} else {
+				try {
+					this.responseText = FS.readFileSync(decodeURI(url.pathname.slice(1)), 'utf8');
+					this.status = 200;
+					this._setState(XMLHttpRequestReadyStateEnum.done);
+				} catch (error) {
+					this._handleError(error);
+				}
+			}
+
+			return;
+		}
```

Comment:
We need to deal with security issues carefully.
This may leads to arbitrary file reading.

Reply:
hi @capricorn86, do you have any idea to solve it?
I have a idea：
- Add unsafe mode for users to manually choose to enable, just like jsdom.